### ISSUE TITLE
Downloading controlled vocabularies and the data model as efficiently as possible

### ISSUE THREAD

**Issue by moonshoes87** [PmagPy/PmagPy#93]
I'm planning to overwrite some of the changes here, @Caoimhinmg :

made vocabulary a class variable not an instance variable of Contribution so that it doesn't have to import vocab every time it is instanciated, reworked the data_model set process for demag GUI.

https://github.com/PmagPy/PmagPy/commit/ce3b1b674a253ad0a8f1cc60a82f9ed2be6610bb

The reason I set it up the way I did was so that you could open a GUI and have a contribution without necessarily having to connect to Earthref and download a bunch of vocabulary first (in Pmag & Magic GUIs).  This can be slow with a bad internet connection and I preferred to move the delay of downloading to a bit later in the user experience.  

The way I handled the problem that you are trying to fix is by passing in the Vocabulary object during creation of the Contribution (or whatever object depends on it) if one had already been created earlier in this session:

    def __init__(self, directory=".", read_tables='all',
                 custom_filenames=None, single_file=None,
                 dmodel=None, vocabulary=""):

I.e., if I were making my second Contribution I could use the first one's vocabulary to instantiate the second.

Another issue is that I don't want Demag GUI to re-download the vocabulary if it is being opened via Pmag GUI, for example, which also downloads controlled vocabulary.  I'm going to try soon to really go through and make a systematic, clean solution so that vocabularies, data models, etc. are shared easily and cleanly between concerns.  I just wanted to explain this issue a bit more so we aren't working at cross purposes.  I'm planning to make it as efficient as possible from the user perspective, and more clean and obvious from the development side (which I realize it hasn't been so far).

**Comment by Caoimhinmg** [comment]
Then it might be better to do it like this:

```python
    def __init__(self, directory=".", read_tables='all',
                 custom_filenames=None, single_file=None,
                 dmodel=None, vocabulary=""):
        if len(vocabulary):
            self.cv = vocabulary
        else:
            try: self.cv = Contribution.cv
            except AttributeError:
                vocab = cv.Vocabulary()
                Contribution.cv, _ = vocab.get_controlled_vocabularies()
                self.cv = Contribution.cv
```

This way you can pass in vocabulary, or just call it and not have it reimport. This way all Contribution instances will share vocabulary and not reimport unless told to and at any time you can overwrite the instance of vocabulary for a single Contribution by just passing in a vocabulary in __init__. I won't have to find every instantiation of Contribution and pass in a vocab, and you won't have the vocab imported on program start. What do you think @moonshoes87 ?

**Comment by moonshoes87** [comment]
A solution along these lines can work, but I need to do some more tweaking to prevent redundant imports in other places (like the grid_frame module), and get the timing right for Pmag and MagIC GUIs so that they don’t take too long to start (the current solution does create the delay I’m trying to avoid).  

I’ll work on it a bit more and let you know what I come to.

**Comment by moonshoes87** [comment]
So, unfortunately I made a bunch of changes that I wasn’t quite able to push before going to the hospital.  Mostly these changes were to fix how the controlled vocabularies are downloaded.  Kevin, I’m going overwrite your changes for how the vocabulary is imported a little bit.  Similar idea, but instead of Contribution.cv being the list of controlled vocabulary, it will be the Vocabulary class object, which has a few more useful things in it (method codes, etc.).  

I will go through the code to make sure that this works correctly in all of the GUIs, and that they expect the correct Vocabulary object.

I’m dealing with an irritating rebase right now, so please don’t mess with any of the controlled vocabulary stuff in the meantime.  Thanks!

**Comment by moonshoes87** [comment]
It looks like all of this is working correctly now.  

The Contribution object will now have as an attribute the higher level Vocabulary (it used to just have Vocabulary.vocabularies, which is a Series of the controlled vocabularies).  The higher level Vocabulary has sorted method codes and some other attributes.  

I went through the code to make sure that in all instances a Vocabulary object is what is passed and expected.  I believe I fixed it everywhere that needed fixing, but there could be something buried in Demag GUI code that uses controlled vocabularies that I didn't find.  @Swanson-Hysell and @Caoimhinmg, if you want to verify that I haven't missed any controlled vocabularies in your code that would be great.  Thanks!

**Comment by Caoimhinmg** [comment]
@moonshoes87 looks good to me. I've tested importing without parameters and checking that the same exact object in memory is the vocabulary unless otherwise specified by a init parameter and it seems to be working. It's only importing once when the GUIs run which was the main runtime advantage of this method so looks good to me. Also sorry about any problems introduced into the controlled vocabulary read in functions I had to make some of the changes because of errors I kept getting in the binaries I was trying to get done by around this time.
